src/merkle.rs

```diff
@@ -1,51 +1,182 @@
-use ff::{PrimeField, BitIterator};
+use std::ops::{Deref, DerefMut};
+use ff::{PrimeField, BitIterator, Field};
 use pairing::bls12_381::Fr;
 use zcash_primitives::jubjub::JubjubEngine;
 use zcash_primitives::pedersen_hash;
+use crate::Params;
 
-#[derive(Debug, Clone, Copy, Eq, PartialEq)]
-pub enum PathDirection {
-    Left,
-    Right,
+/// A point in the authentication path.
+#[derive(Clone, Debug)]
+pub struct AuthPathPoint<E: JubjubEngine> {
+    /// The current selection. That is, the opposite of sibling.
+    pub current_selection: MerkleSelection,
+    /// Sibling value, if it exists.
+    pub sibling: Option<E::Fr>,
 }
 
-impl PathDirection {
-    pub fn random<R: rand_core::RngCore>(rng: &mut R) -> Self {
-        if rng.next_u32() % 2 == 0 {
-            PathDirection::Left
-        } else {
-            PathDirection::Right
+/// The authentication path of the merkle tree.
+#[derive(Clone, Debug)]
+pub struct AuthPath<E: JubjubEngine>(pub Vec<AuthPathPoint<E>>);
+
+impl<E: JubjubEngine> AuthPath<E> {
+    /// Create a random path.
+    pub fn random<R: rand_core::RngCore>(depth: usize, rng: &mut R) -> Self {
+        Self(vec![AuthPathPoint {
+            current_selection: MerkleSelection::random(rng),
+            sibling: Some(<E::Fr>::random(rng))
+        }; depth])
+    }
```

Comment:
Should we feature gate `AuthPath::random` and `MerkleSelection::random` on `test`?